### PR TITLE
feat(S3): add S3 service and checks

### PR DESCRIPTION
**Pull request by MrCloudSec**
### Context 

Add S3 service and checks.

### Description

- Add all s3 tests.
-  Add the following checks:

checkName | oldCheck
-- | --
s3_bucket_public_access | extra73
s3_bucket_default_encryption | extra734
s3_bucket_secure_transport_policy | extra764
s3_bucket_acl_prohibited | extra7172
cloudtrail_s3_dataevents_enabled | extra725
s3_bucket_policy_public_write_access | extra771

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled.metadata.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  "SubServiceName": "",
+  "ResourceIdTemplate": "arn:partition:service:region:account-id:resource-id",
+  "Severity": "low",
+  "ResourceType": "AwsS3Bucket",
+  "Description": "Ensure that all your AWS CloudTrail trails are configured to log Data events in order to record S3 object-level API operations, such as GetObject, DeleteObject and PutObject, for individual S3 buckets or for all current and future S3 buckets provisioned in your AWS account.",
+  "Risk": "If logs are not enabled, monitoring of service use and threat analysis is not possible.",
+  "RelatedUrl": "",
+  "Remediation": {
+    "Code": {
+      "CLI": "aws cloudtrail put-event-selectors --trail-name cc-global-env-trail --event-selectors '[{ 'ReadWriteType': 'All', 'IncludeManagementEvents':true, 'DataResources': [{ 'Type': 'AWS::S3::Object', 'Values': ['arn:aws:s3'] }] }]'",
```
Comment (jfagoagas) [inline providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled.metadata.json]:
```suggestion
      "CLI": "aws cloudtrail put-event-selectors --trail-name <YOUR_TRAIL_NAME_HERE> --event-selectors '[{ 'ReadWriteType': 'All', 'IncludeManagementEvents':true, 'DataResources': [{ 'Type': 'AWS::S3::Object', 'Values': ['arn:aws:s3'] }] }]'",
```
What do you think? I think it's better to understand.

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            result = check.execute()
+
+            assert len(result) == 1
+            assert result[0].status == "FAIL"
+            assert search(
+                "No CloudTrail trails have a data event to record all S3 object-level API operations.",
+                result[0].status_extended,
+            )
+            assert result[0].resource_id == "No trails"
+            assert result[0].resource_arn == "No trails"
```
Comment (jfagoagas) [inline providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:45]: Include region validation here pls.

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
```
Comment (jfagoagas) [inline providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:94]: Include region validation here pls.

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            # Test Check
+            from providers.aws.services.cloudtrail.cloudtrail_s3_dataevents_enabled.cloudtrail_s3_dataevents_enabled import (
+                cloudtrail_s3_dataevents_enabled,
+            )
+
+            check = cloudtrail_s3_dataevents_enabled()
+            result = check.execute()
+
+            assert len(result) == 1
+            assert result[0].status == "PASS"
```
Comment (jfagoagas) [inline providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:137]: Include region validation here pls.

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            check = s3_bucket_acl_prohibited()
+            result = check.execute()
+
+            assert len(result) == 1
+            assert result[0].status == "FAIL"
+            assert search(
+                "ACLs enabled",
+                result[0].status_extended,
+            )
+            assert result[0].resource_id == bucket_name_us
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:38]: Please include region here, thanks!

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            check = s3_bucket_acl_prohibited()
+            result = check.execute()
+
+            assert len(result) == 1
+            assert result[0].status == "PASS"
+            assert search(
+                "ACLs disabled",
+                result[0].status_extended,
+            )
+            assert result[0].resource_id == bucket_name_us
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:104]: Please include region here, thanks!

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+from re import search
+from unittest import mock
+
+from boto3 import client
+from moto import mock_s3
+
+
+class Test_s3_bucket_acl_prohibited:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:8]: Can we include a test without `BucketOwnerEnforced` in the `bucket.ownership`?

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+class s3_bucket_default_encryption(Check):
+    def execute(self):
+        findings = []
+        for bucket in s3_client.buckets:
+            report = Check_Report(self.metadata)
+            report.region = bucket.region
+            report.resource_id = bucket.name
+            if not bucket.encryption:
+                report.status = "FAIL"
+                report.status_extended = f"Server Side Encryption configuration is not configured for S3 Bucket {bucket.name}."
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py]:
```suggestion
                report.status_extended = f"Server Side Encryption is not configured for S3 Bucket {bucket.name}."
```

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            else:
+                # Check if bucket policy enforce SSE
+                if not bucket.policy:
+                    report.status = "FAIL"
+                    report.status_extended = f"S3 Bucket {bucket.name} has default encryption with {bucket.encryption} but does not have a bucket policy."
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py]: Why this is a FAIL?

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            check = s3_bucket_default_encryption()
+            result = check.execute()
+
+            assert len(result) == 1
+            assert result[0].status == "FAIL"
+            assert search(
+                "Server Side Encryption configuration is not configured",
+                result[0].status_extended,
+            )
+            assert result[0].resource_id == bucket_name_us
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py:38]: Please include the region here

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                        "KMSMasterKeyID": "12345678",
+                    }
+                }
+            ]
+        }
+
+        s3_client_us_east_1.put_bucket_encryption(
+            Bucket=bucket_name_us, ServerSideEncryptionConfiguration=sse_config
+        )
+        encryption_policy = '{"Version": "2012-10-17","Id": "PutObjPolicy","Statement": [{"Sid": "DenyIncorrectEncryptionHeader","Effect": "Deny","Principal": "*","Action": "s3:PutObject","Resource": "arn:aws:s3:::bucket_test_us/*","Condition": {"StringNotEquals": {"s3:x-amz-server-side-encryption": "aws:kms"}}}]}'
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py]:
```suggestion
        encryption_policy = """
{
  "Version": "2012-10-17",
  "Id": "PutObjPolicy",
  "Statement": [
    {
      "Sid": "DenyIncorrectEncryptionHeader",
      "Effect": "Deny",
      "Principal": "*",
      "Action": "s3:PutObject",
      "Resource": "arn:aws:s3:::bucket_test_us/*",
      "Condition": {
        "StringNotEquals": {
          "s3:x-amz-server-side-encryption": "aws:kms"
        }
      }
    }
  ]
}
""" 
```

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
+from re import search
+from unittest import mock
+
+from boto3 import client
+from moto import mock_s3
+
+
+class Test_s3_bucket_default_encryption:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py:8]: Are you testing all the possible scenarios for the bucket encryption?

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        ):
+            # Test Check
+            from providers.aws.services.s3.s3_bucket_default_encryption.s3_bucket_default_encryption import (
+                s3_bucket_default_encryption,
+            )
+
+            check = s3_bucket_default_encryption()
+            result = check.execute()
+
+            assert len(result) == 1
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py]: Please include the region here

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py:79]: Please include the region here

File: providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning.metadata.json [hunk 15]
```diff
@@ -15,8 +15,8 @@
     "Code": {
       "CLI": "",
       "NativeIaC": "",
-      "Other": "",
-      "Terraform": ""
+      "Other": "https://docs.bridgecrew.io/docs/s3_16-enable-versioning#aws-console",
+      "Terraform": "https://docs.bridgecrew.io/docs/s3_16-enable-versioning#terraform"
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning.metadata.json:19]: 🔝

File: providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning_test.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
+from re import search
+from unittest import mock
+
+from boto3 import client
+from moto import mock_s3
+
+
+class Test_s3_bucket_object_versioning:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning_test.py:8]: Please include the region here

File: providers/aws/services/s3/s3_bucket_policy_public_write_access/s3_bucket_policy_public_write_access_test.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
+from re import search
+from unittest import mock
+
+from boto3 import client
+from moto import mock_s3
+
+
+class Test_s3_bucket_policy_public_write_access:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_policy_public_write_access/s3_bucket_policy_public_write_access_test.py:8]: Please include the region here

File: providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        # 1. Check if public buckets are restricted at account level
+        if (
+            s3control_client.account_public_access_block.ignore_public_acls
+            and s3control_client.account_public_access_block.restrict_public_buckets
+        ):
+            report = Check_Report(self.metadata)
+            report.status = "PASS"
+            report.status_extended = f"All S3 public access blocked at account level."
+            report.region = s3control_client.region
+            report.resource_id = ""
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py]:
```suggestion
            report.resource_id = s3_client.audited_account
```

File: providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access_test.py [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+from boto3 import client, session
+from moto import mock_s3, mock_s3control
+
+from providers.aws.lib.audit_info.models import AWS_Audit_Info
+
+AWS_ACCOUNT_NUMBER = "123456789012"
+AWS_REGION = "us-east-1"
+
+
+class Test_s3_bucket_public_access:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access_test.py:13]: Can you include a test with the account level block and no buckets?

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        findings = []
+        for bucket in s3_client.buckets:
+            report = Check_Report(self.metadata)
+            report.region = bucket.region
+            report.resource_id = bucket.name
+            # Check if bucket policy enforces SSL
+            if not bucket.policy:
+                report.status = "FAIL"
+                report.status_extended = (
+                    f"S3 Bucket {bucket.name} does not have a bucket policy."
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy.py]:
```suggestion
                    f"S3 Bucket {bucket.name} does not have a bucket policy, therefore allows requests without SSL/TLS."
```

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
+from re import search
+from unittest import mock
+
+from boto3 import client
+from moto import mock_s3
+
+
+class Test_s3_bucket_secure_transport_policy:
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py:8]: Include region assert pls

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+            )
+            assert result[0].resource_id == bucket_name_us
+
+    @mock_s3
+    def test_bucket_comply_policy(self):
+        s3_client_us_east_1 = client("s3", region_name="us-east-1")
+        bucket_name_us = "bucket_test_us"
+        s3_client_us_east_1.create_bucket(Bucket=bucket_name_us)
+
+        ssl_policy = '{"Version": "2012-10-17","Id": "PutObjPolicy","Statement": [{"Sid": "s3-bucket-ssl-requests-only","Effect": "Deny","Principal": "*","Action": "s3:PutObject","Resource": "arn:aws:s3:::bucket_test_us/*","Condition": {"Bool": {"aws:SecureTransport": "false"}}}]}'
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py]:
```suggestion
        ssl_policy = """
{
  "Version": "2012-10-17",
  "Id": "PutObjPolicy",
  "Statement": [
    {
      "Sid": "s3-bucket-ssl-requests-only",
      "Effect": "Deny",
      "Principal": "*",
      "Action": "s3:PutObject",
      "Resource": "arn:aws:s3:::bucket_test_us/*",
      "Condition": {
        "Bool": {
          "aws:SecureTransport": "false"
        }
      }
    }
  ]
}
"""        
```

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+            )
+            assert result[0].resource_id == bucket_name_us
+
+    @mock_s3
+    def test_bucket_uncomply_policy(self):
+        s3_client_us_east_1 = client("s3", region_name="us-east-1")
+        bucket_name_us = "bucket_test_us"
+        s3_client_us_east_1.create_bucket(Bucket=bucket_name_us)
+
+        ssl_policy = '{"Version": "2012-10-17","Id": "PutObjPolicy","Statement": [{"Sid": "s3-bucket-ssl-requests-only","Effect": "Deny","Principal": "*","Action": "s3:GetObject","Resource": "arn:aws:s3:::bucket_test_us/*","Condition": {"Bool": {"aws:SecureTransport": "false"}}}]}'
```
Comment (jfagoagas) [inline providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py]:
```suggestion
        ssl_policy = """
{
  "Version": "2012-10-17",
  "Id": "PutObjPolicy",
  "Statement": [
    {
      "Sid": "s3-bucket-ssl-requests-only",
      "Effect": "Deny",
      "Principal": "*",
      "Action": "s3:GetObject",
      "Resource": "arn:aws:s3:::bucket_test_us/*",
      "Condition": {
        "Bool": {
          "aws:SecureTransport": "false"
        }
      }
    }
  ]
}
"""        
```

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Impressive!!!!!!! Please @sergargar review my comments to merge this PR

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py (hunk 2 above)
Comment (MrCloudSec) [reply providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:45]: Done

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py (hunk 3 above)
Comment (MrCloudSec) [reply providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:94]: Done

File: providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py (hunk 4 above)
Comment (MrCloudSec) [reply providers/aws/services/cloudtrail/cloudtrail_s3_dataevents_enabled/cloudtrail_s3_dataevents_enabled_test.py:137]: Done

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py (hunk 6 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:104]: Done

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py (hunk 5 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:38]: Done

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py (hunk 9 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py]: You're right, AWS recently changed it and the policy is not necessary:
https://aws.amazon.com/premiumsupport/knowledge-center/bucket-policy-encryption-s3/

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py (hunk 8 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption.py]: Done

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py (hunk 10 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py:38]: Done

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py (hunk 13 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py]: Done

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py (hunk 11 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py]: No longer needed.

File: providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning_test.py (hunk 16 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_object_versioning/s3_bucket_object_versioning_test.py:8]: Done

File: providers/aws/services/s3/s3_bucket_policy_public_write_access/s3_bucket_policy_public_write_access_test.py (hunk 17 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_policy_public_write_access/s3_bucket_policy_public_write_access_test.py:8]: Done

File: providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py (hunk 18 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py]: Done

File: providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access_test.py (hunk 19 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access_test.py:13]: Done

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy.py (hunk 20 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy.py]: Done

File: providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py (hunk 21 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_secure_transport_policy/s3_bucket_secure_transport_policy_test.py:8]: Done

File: providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py (hunk 12 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_default_encryption/s3_bucket_default_encryption_test.py:8]: Yes

File: providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py (hunk 7 above)
Comment (MrCloudSec) [reply providers/aws/services/s3/s3_bucket_acl_prohibited/s3_bucket_acl_prohibited_test.py:8]: Done

**Review by jfagoagas — APPROVED** [review]

**Pull request merged**